### ISSUE TITLE
Let users pass an RNG object to nondeterministic functions

### ISSUE THREAD

**Issue by gregor-robinson** [CitrineInformatics/lolo#206]
For the sake of reproducible results, including deterministic unit and integration testing, Bagger and GuessTheMeanLearner should have access to a user-specified RNG object (at least at training time). It would be possible to expose an interface that allows users to specify a RNG seed, but that is strictly less powerful and prevents the use of desirable alternative RNGs such as those that are counter-based. Any other nondeterministic functionality should similarly accept an RNG object, so please note any other nondeterminism to keep in mind -- cross validation, for example.

**Comment by gregor-robinson** [comment]
Heavy lift would be gained by doing this just for bag creation, which can be standardized in BaggerHelper (see https://github.com/CitrineInformatics/lolo/issues/212)

**Comment by mrupp-citrine** [comment]
👍 for control of pseudo-random numbers. But why is it necessary to pass the generator object? A seed should be fine _as long as all the following results are deterministic_. This would be the case for example if an instance of the NumPy generator would be created locally. That is not the most efficient solution (this might be a very efficient approach: https://github.com/google/jax/blob/master/design_notes/prng.md) but easy to implement.

**Comment by gregor-robinson** [comment]
It's good practice to expose an option to pass a generator rather than a seed because it allows the user to swap the RNG with a different algorithm. This helps make code more testable, instrumentable, and flexible --- and it's negligible implementation cost.

**Comment by mrupp-citrine** [comment]
I would argue that passing a generator increases interface complexity but provides limited benefit except for cases where the effect of the generator itself should be tested. The latter is unlikely to become a priority, and can be (partially) done with specialized test suites such as the diehard tests.

In the unlikely case that passing a generator should become necessary, the seed argument could be upgraded to int or generator, using a default generator if only the seed is specified. So passing a seed does not preclude later passing of generators.

**Comment by bfolie** [comment]
`Bagger`, `GuessTheMeanLearner`, `RegressionTreeLearner`, `ClassificationTreeLearner`, `MultiTaskTree`, `RegressionSplitter`, `ClassificationSplitter`, `MultiTaskSplitter`, `RandomForest`, and `ExtraRandomTrees` all have access to random number generators, so I'm closing this.

I noticed one place where the `rng` is not passed through in `RandomForest`, so I fixed that in #258.

Our random forest still isn't fully reproducible because the trees are trained in parallel. I opened an issue: #259
